File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int[] pre = new int[n];
        int[] post = new int[n];
        for(int i=0; i<n; i++){
            pre[i] = 0;
            post[i] = 0;
        }
        for(int i=1; i<n; i++){
            pre[i] = Math.max(pre[i-1], height[i-1]);
            post[n-i-1] = Math.max(post[n-i], height[n-i]);
        }

        int water = 0;
        for(int i=0; i<n; i++){
            int curr = Math.min(pre[i], post[i]) - height[i];
            if(curr > 0)
                water += curr;
        }
        return water;
    }
}

File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int> pref(n, 0), post(n, 0);
        for(auto i=0; i<n-1; i++){
            pref[i+1] = max(pref[i], height[i]);
            post[n-i-2] = max(post[n-i-1], height[n-i-1]);
        }
        int store = 0;
        for(auto i=0; i<n; i++){
            int water = min(pref[i], post[i]) - height[i];
            if(water > 0)
                store += water;
        }
        return store;
    }
};